### ISSUE TITLE
Add a `soft_assert` that a query's result type is not changed during optimization

### ISSUE THREAD

**Issue by ggevay** [MaterializeInc/materialize#27326]
Optimizations shouldn't change a query's result type in incompatible ways. (Stuff like refining nullability or keys is allowed, but e.g. changing from `text` to `int` is not.) Checking this could catch some bugs.

We discussed this a few times before. It most recently came up here: https://github.com/MaterializeInc/materialize/pull/27300#discussion_r1616041225

cc @mgree, @aalexandrov

**Comment by mgree** [comment]
Yes, good idea! This seems like an easy thing to do that would help catch a serious bug.
